pkg/services/secrets/kvstore/migrations/migrator.go

```diff
@@ -0,0 +1,61 @@
+package migrations
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"reflect"
+
+	"github.com/grafana/grafana/pkg/infra/log"
+)
+
+var logger = log.New("secret.migration")
+
+// SecretMigrationServiceProvider provides SecretMigration services.
+type SecretMigrationServiceProvider interface {
+	GetServices() []SecretMigrationService
+	Run(context.Context) error
+}
+
+// SecretMigrationService is used to migrate legacy secrets to new unified secrets.
+type SecretMigrationService interface {
+	Run(ctx context.Context) error
+}
+
+type SecretMigrationServiceProviderImpl struct {
+	Services []SecretMigrationService
+}
+
+func ProvideSecretMigrationServiceProvider(
+	dataSourceSecretMigrationService *DataSourceSecretMigrationService,
+) *SecretMigrationServiceProviderImpl {
+	return NewSecretMigrationServiceProvider(
+		dataSourceSecretMigrationService,
+	)
+}
+
+func NewSecretMigrationServiceProvider(services ...SecretMigrationService) *SecretMigrationServiceProviderImpl {
+	return &SecretMigrationServiceProviderImpl{services}
+}
+
+func (s *SecretMigrationServiceProviderImpl) GetServices() []SecretMigrationService {
+	return s.Services
+}
+
+// Run migration services. This will block until all services have exited.
+func (s *SecretMigrationServiceProviderImpl) Run(ctx context.Context) error {
+	services := s.Services
+
+	// Start migration services.
+	for _, service := range services {
+		serviceName := reflect.TypeOf(service).String()
+		logger.Debug("Starting secret migration service", "service", serviceName)
+		err := service.Run(ctx)
+		if err != nil && !errors.Is(err, context.Canceled) {
+			logger.Error("Stopped secret migration service", "service", serviceName, "reason", err)
+			return fmt.Errorf("%s run error: %w", serviceName, err)
+		}
+		logger.Debug("Finished secret migration service", "service", serviceName)
```

Comment:
Oh, will add a check for that, didn't realize that could happen, learning more about Go context everyday.

Reply:
Should I just removed the `!errors.Is(err, context.Canceled) {` from the err check on line 45?

Reply:
Removed on https://github.com/grafana/grafana/pull/50463/commits/9ad12e62d8a779d55ce2879d9900751d3a578f63 since it should just return an error to server if the migration is interrupted.